engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/result/NadelResultTransformer.kt

```diff
@@ -31,32 +35,85 @@ internal class NadelResultTransformer(private val executionBlueprint: NadelOvera
+
+        suspend fun getAsync(): List<NadelResultInstruction> {
+            val deferredInstructions = ArrayList<Deferred<List<NadelResultInstruction>>>()
+
+            coroutineScope {
+                for ((field, steps) in executionPlan.transformationSteps) {
+                    // This can be null if we did not end up sending the field e.g. for hydration
+                    val underlyingFields = overallToUnderlyingFields[field]
+                    if (underlyingFields == null || underlyingFields.isEmpty()) {
+                        continue
+                    }
+
+                    for (step in steps) {
+                        deferredInstructions.add(
+                            async {
+                                step.transform.getResultInstructions(
+                                    executionContext,
+                                    executionBlueprint,
+                                    service,
+                                    field,
+                                    underlyingFields.first().parent,
+                                    result,
+                                    step.state,
+                                    nodes,
+                                )
+                            },
+                        )
+                    }
+                }
+            }
+
+            return deferredInstructions
+                .awaitAll()
+                .flatten()
+        }
+
+        suspend fun getSync(): List<NadelResultInstruction> {
+            val instructions = ArrayList<NadelResultInstruction>()
+
+            for ((field, steps) in executionPlan.transformationSteps) {
                 // This can be null if we did not end up sending the field e.g. for hydration
                 val underlyingFields = overallToUnderlyingFields[field]
-                    ?.takeIf(List<ExecutableNormalizedField>::isNotEmpty)
-                    ?: return@step emptyList()
-
-                step.transform.getResultInstructions(
-                    executionContext,
-                    executionBlueprint,
-                    service,
-                    field,
-                    underlyingFields.first().parent,
-                    result,
-                    step.state,
-                )
+                if (underlyingFields == null || underlyingFields.isEmpty()) {
+                    continue
+                }
+
+                for (step in steps) {
+                    instructions.addAll(
+                        step.transform.getResultInstructions(
+                            executionContext,
+                            executionBlueprint,
+                            service,
+                            field,
+                            underlyingFields.first().parent,
+                            result,
+                            step.state,
+                            nodes,
+                        ),
+                    )
+                }
```

Comment:
The innards of this sync / async code looks the same no?  We should extract it into a common function versus repeating it.

Reply:
I mean, yes but hopefully the FF gets deleted after it's proven that it works? I don't think we'll maintain this code for long.

Reply:
fair point